File: server/lib/middleware/api_enabled_middleware.dart
import 'dart:convert';

import 'package:shelf/shelf.dart';

import '../services/plugin_data_service.dart';

/// API 启用检查中间件
///
/// 检查用户是否已设置加密密钥
/// 加密密钥通过 X-Encryption-Key 请求头传递
Middleware apiEnabledMiddleware(PluginDataService pluginDataService) {
  return (Handler innerHandler) {
    return (Request request) async {
      final userId = request.context['userId'] as String?;

      if (userId == null) {
        return _errorResponse(401, '未认证');
      }

      // 从请求头获取加密密钥
      final encryptionKey = request.headers['x-encryption-key'];

      if (encryptionKey == null || encryptionKey.isEmpty) {
        return _errorResponse(
          403,
          '请通过 X-Encryption-Key 请求头传递加密密钥',
        );
      }

      // 设置加密密钥到内存
      pluginDataService.setEncryptionKey(userId, encryptionKey);

      // 将加密密钥添加到请求上下文
      final updatedRequest = request.change(
        context: {
          ...request.context,
          'encryptionKey': encryptionKey,
        },
      );

      return innerHandler(updatedRequest);
    };
  };
}

Response _errorResponse(int statusCode, String message) {
  return Response(
    statusCode,
    body: jsonEncode({
      'success': false,
      'error': message,
      'timestamp': DateTime.now().toIso8601String(),
    }),
    headers: {'Content-Type': 'application/json'},
  );
}

File: server-nodejs/src/middleware/apiEnabledMiddleware.ts
import { Request, Response, NextFunction } from 'express';

/**
 * API 启用中间件
 *
 * 检查请求头中是否包含加密密钥（用于插件 API 访问）
 * 密钥通过 X-Encryption-Key 请求头传递
 */
export function apiEnabledMiddleware() {
  return (req: Request, res: Response, next: NextFunction): void => {
    const userId = req.context?.userId;

    if (!userId) {
      res.status(401).json({
        success: false,
        error: '未认证',
        timestamp: new Date().toISOString(),
      });
      return;
    }

    // 从请求头获取加密密钥
    const encryptionKey = req.headers['x-encryption-key'] as string | undefined;
    if (!encryptionKey || encryptionKey.length === 0) {
      res.status(403).json({
        success: false,
        error: '请通过 X-Encryption-Key 请求头传递加密密钥',
        timestamp: new Date().toISOString(),
      });
      return;
    }

    // 将加密密钥存储到请求上下文中
    req.context = req.context || {};
    req.context.encryptionKey = encryptionKey;

    next();
  };
}
